protocols/protocol/cloudstate/event_sourced.proto

```diff
@@ -81,35 +123,70 @@ message EventSourcedReply {
+//    - `message` MUST be one of the following:
+//      - `init` TODO
+//      - `event` TODO
+//      - `command` TODO
 message EventSourcedStreamIn {
+
     oneof message {
+        // TODO
         EventSourcedInit init = 1;
+        // TODO
         EventSourcedEvent event = 2;
+        // TODO
         Command command = 3;
     }
 }
 
-// Output message type for the gRPC stream out.
+//*EventSourcedStreamOut* TODO
+//
+//---
+//
+//`Specification`:
+//
+//  - An `EventSourcedStreamOut` is considered to be valid iff:
+//    - `message` MUST be one of the following:
+//      - `reply` TODO
+//      - `failure` TODO
 message EventSourcedStreamOut {
+
     oneof message {
+        // TODO
         EventSourcedReply reply = 1;
+        // TODO
         Failure failure = 2;
     }
 }
 
-// The Entity service
+//The *EventSourced* Service is the entrypoint for event sourced User Function entities.
+//
+// TODO: One stream will be established per active entity.
+// Once established, the first message sent will be Init, which contains the entity ID, and,
+// if the entity has previously persisted a snapshot, it will contain that snapshot. It will
+// then send zero to many event messages, one for each event previously persisted. The entity
+// is expected to apply these to its state in a deterministic fashion. Once all the events
+// are sent, one to many commands are sent, with new commands being sent as new requests for
+// the entity come in. The entity is expected to reply to each command with exactly one reply
+// message. The entity should reply in order, and any events that the entity requests to be
+// persisted the entity should handle itself, applying them to its own state, as if they had
+// arrived as events when the event stream was being replayed on load.
+//
+//---
+//
+//`Specification`:
+//
+//  - An implementation of `EventSourced` is only valid iff:
+//  - `handle`
+//    - TODO
 service EventSourced {
-
-    // The stream. One stream will be established per active entity.
-    // Once established, the first message sent will be Init, which contains the entity ID, and,
-    // if the entity has previously persisted a snapshot, it will contain that snapshot. It will
-    // then send zero to many event messages, one for each event previously persisted. The entity
-    // is expected to apply these to its state in a deterministic fashion. Once all the events
-    // are sent, one to many commands are sent, with new commands being sent as new requests for
-    // the entity come in. The entity is expected to reply to each command with exactly one reply
-    // message. The entity should reply in order, and any events that the entity requests to be
-    // persisted the entity should handle itself, applying them to its own state, as if they had
-    // arrived as events when the event stream was being replayed on load.
+    // TODO
     rpc handle(stream EventSourcedStreamIn) returns (stream EventSourcedStreamOut) {}
```

Comment:
Beside sending a `protocol.Failure` and then closing the stream, is a user function allowed to close a stream without a failure sent before and the user function indicating a non-failure reason for the stream closed by a [gRPC status code](https://grpc.github.io/grpc/cpp/namespacegrpc.html#aff1730578c90160528f6a8d67ef5c43b)?

Reply:
I'd say it SHOULD NOT do this, as opposed to MUST NOT (rfc2119 notation). There's always going to be situations where the user function may close the stream without sending a failure - eg, if the process crashes. But the problem with allowing such a scenario is if the proxy has a command to send to it, just before it receives the stream closed signal, that command is now going to fail when it tries to send. It's the proxy that decides when entities are active or not.

If we wanted, we could introduce a way for user functions to initiate the passivation of entities, that's pretty straight forward, however, closing the stream won't cut it, first the user function has to send a message to request the proxy to passivate, then the proxy can start its passivation sequence, which is to first drain its queue of existing commands, then start buffering any new commands that come in, and then passivate the entity, and finally if any new commands did come in while the entity was being passivated, restarting it to handle those buffered commands. Simply closing the stream is not enough.